---
title: "Experts session: Hands-on Ribo-Seq workflows - RiboViz"
author: "Riboviz Team"
date: "02/07/2021"
output:
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/..."></footer>');
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 100px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>

<div class="notes">

Bio:

Title: Experts session: Hands-on Ribo-Seq workflows - RiboViz

Abstract:

Overview of riboviz workflow and features
Hands-on session to run riboviz on a small dataset
 - Run on vignette
 - Run on downsampled dataset
 - Finding and interpreting riboviz output
Overview of how to adapt to your own dataset
Question-and-answer session
Evaluation

</div>

# Hands-on Ribo-Seq workflows - RiboViz


<div class="notes">

...

</div>

## Outline:

* Overview of riboviz workflow and features
* Hands-on session to run riboviz on a small dataset
  - Run on vignette
  - Run on downsampled dataset
  - Finding and interpreting riboviz output
* Overview of how to adapt to your own dataset
* Question-and-answer session
* Evaluation



# Overview of riboviz workflow and features

<div class="notes">

Short slide presentation that includes:
 - goals of this workshop session
 - goals of riboviz
 - mention key design decisions
    - nextflow
    - transcriptome-centric 
    - h5 file 
    - standard file outputs of most other things (bam, bedgraph, flat .txt) 
    - apologize that we do not currently have RNA-seq in parallel
 - overview of the workflow, i.e. what happens
 - describe key ingredients 
    - data in fastq format
    - annotation files 
    - configuration file 
    - mention that we can deal with multiplexed input but do not get into detail

Next: move to hands-on

</div>

## riboviz workflow 

<div class="notes">
... slide notes ...
</div>

## riboviz features 

<div class="notes">
... slide notes ...
</div>



# Hands-on session to run riboviz on a small dataset

## Run on vignette  

<div class="notes">
This should show what it "looks like" for riboviz to run and take about 3 minutes.

We can flush out bugs or installation problems.

We can talk through what the nextflow log means, where the output goes, and what the html output means, very briefly.
</div>

## Run on downsampled dataset  

<div class="notes">
Hands-on approach to:
 - download the data + config file
 - ensure that prep_riboviz.nf can find these inputs, using --validate-only
 - run on that data, ideally takes under 5 minutes on a laptop.
 - as the data are running, Q&A on dataset setup and nextflow
 - refer to documentation, especially troubleshooting
</div>

## Finding and interpreting riboviz output  

<div class="notes">
 - talk through where the files are
 - show and talk through .html output from the new downsampled dataset.
 - have one example from another full-size dataset to compare?
</div>



# Overview of how to adapt to your own dataset

<div class="notes">
focus on key considerations: 
 - annotation files (inc. contaminants) - checkfastagff tool
 - the annotation file follows the biological question and must be chosen by the user
 - read structure (adapters, UMI/regex)
 - everything goes in config.yaml
refer to documentation for config.yaml
explain example-datasets repository
</div>



# Questions & Answer Session

<div class="notes">
... slide notes ...
</div>



# Workshop Evaluation

<div class="notes">
... slide notes ...
</div>